src/dmd/cparse.d

```diff
@@ -3671,60 +3671,166 @@ final class CParser(AST) : Parser!AST
+                // if we've already seen ( type-name ), then this is a cast
+                break;
+        }
+        pt = t;
+        return true;
+    }
 
+    /********************************
+     * See if match for unary-expression.
+     * Params:
+     *    pt = starting token, updated to one past end of constant-expression if true
+     *    afterParenType = true if already seen ( type-name ) of a cast-expression
+     * Returns:
+     *    true if unary-expression
+     */
+    private bool isUnaryExpression(ref Token* pt, bool afterParenType = false)
+    {
+        auto t = pt;
         switch (t.value)
         {
-        case TOK.leftParenthesis:
-        case TOK.plusPlus:
-        case TOK.minusMinus:
-        case TOK.not:
-        case TOK.tilde:
-        case TOK.sizeof_:
-        case TOK._Alignof:
-        case TOK.identifier:
-            return true;
+            case TOK.plusPlus:
+            case TOK.minusMinus:
+                t = peek(t);
+                if (!isUnaryExpression(t, afterParenType))
+                    return false;
+                break;
 
-        case TOK.leftCurly:
-            return true;  // ( type-name ) { initializer-list }
+            case TOK.and:
+            case TOK.mul:
+            case TOK.min:
+            case TOK.add:
+            case TOK.not:
+            case TOK.tilde:
+                t = peek(t);
+                if (!isCastExpression(t, afterParenType))
+                    return false;
+                break;
 
-        case TOK.and:
-        case TOK.mul:
-        case TOK.min:
-        case TOK.add:
-            /* maybe, but just assume it is because (identifer)*unary-expresion
-             * is an expression, but has no effect and so assume it isn't one
-             */
-            return true;
+            case TOK.sizeof_:
+                t = peek(t);
+                if (t.value == TOK.leftParenthesis)
+                {
+                    auto tk = peek(t);
+                    if (isTypeName(tk))
+                    {
+                        if (tk.value != TOK.rightParenthesis)
+                            return false;
+                        t = peek(tk);
+                        break;
+                    }
+                }
+                if (!isUnaryExpression(t, afterParenType))
+                    return false;
+                break;
 
-        default:
-            return false; // ( expression )
+            case TOK._Alignof:
+                t = peek(t);
+                if (t.value != TOK.leftParenthesis)
+                    return false;
+                t = peek(t);
+                if (!isTypeName(t) || t.value != TOK.rightParenthesis)
+                    return false;
+                break;
+
+            default:
+                // Compound literals are handled by cast and sizeof expressions,
+                // so be content with just seeing a primary expression.
+                if (!isPrimaryExpression(t))
```

Comment:
What I mean here is that I deliberately ignore postfix operators `[]` `()` `++` `--` `.` `->`.  Would it be better to be more clear here?